modules/osdk-monitoring-custom-metrics.adoc

```diff
@@ -0,0 +1,158 @@
+// Module included in the following assemblies:
+//
+// * operators/operator_sdk/osdk-monitoring-prometheus.adoc
+
+[id="osdk-monitoring-custom-metrics_{context}"]
+= Exposing custom metrics
+
+As an Operator author, you can publish custom metrics by using the global Prometheus registry from the `controller-runtime/pkg/metrics` library.
+
+.Prerequisites
+
+* Go-based Operator generated using the Operator SDK
+* {product-title} cluster, which deploys the Prometheus Operator by default
+
+.Procedure
+
+. In your Operator SDK project, uncomment the following line in the `config/default/kustomization.yaml` file:
++
+[source,yaml]
+----
+../prometheus
+----
+
+. Create a custom controller class to publish additional metrics from the Operator. The following example declares the `widgets` and `widgetFailures` collectors as global variables, and then registers them with the `init()` function in the controller's package:
++
+.`controllers/memcached_controller_test_metrics.go` file
+[%collapsible]
+====
+[source,go]
+----
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+
+var (
+    widgets = prometheus.NewCounter(
+        prometheus.CounterOpts{
+            Name: "widgets_total",
+            Help: "Number of widgets processed",
+        },
+    )
+    widgetFailures = prometheus.NewCounter(
+        prometheus.CounterOpts{
+            Name: "widget_failures_total",
+            Help: "Number of failed widgets",
+        },
+    )
+)
+
+func init() {
+    // Register custom metrics with the global prometheus registry
+    metrics.Registry.MustRegister(widgets, widgetFailures)
+}
+----
+====
+
+. Record to these collectors from any part of the reconcile loop in the `main` controller class, which determines the business logic for the metric:
++
+.`controllers/memcached_controller.go` file
+[%collapsible]
+====
+[source,go]
+----
+func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	...
+	...
+	// Add metrics
+	widgets.Inc()
+	widgetFailures.Inc()
+
+	return ctrl.Result{}, nil
+}
+----
+====
+
+. Deploy the Operator:
++
+[source,terminal]
+----
+$ make deploy
+----
+
+. Create role and role binding definitions to allow the service monitor of the Operator to be scraped by the Prometheus instance of the {product-title} cluster.
++
+Roles must be assigned so that service accounts have the permissions to scrape the metrics of the namespace:
++
+.`config/prometheus/role.yaml` role
+[%collapsible]
+====
+[source,yaml]
+----
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: prometheus-k8s-role
+  namespace: <operator_namespace>
+rules:
+  - apiGroups:
+      - ""
+    resources:
+      - endpoints
+      - pods
+      - services:q!
```

Comment:
Does this belong? Or were you just trying to get out of Vi?

Reply:
lol wow, I'm going to assume an errant vim escape :sweat_smile: